docs/rholang/rholangmatchingtut.md

```diff
@@ -0,0 +1,181 @@
+
+In the RHO calculus we don't have to worry about distinguishing the top level from other parts of a channel, but because of things like pattern-matching, we have to in Rholang. This will be relevant later on, where there are restrictions on pattern-matching because of this.
+
+## "Looking through the looking glass": Patterns Within Patterns
+
+Patterns follow the same rules as channels in terms of equivalence. In the case of -arity matching, joins, logical connectives, etc. the rules apply to each name pattern individually.
+
+Here we need to treat statements that are not on the top level. We reemphasize that the rules for name equivalences described in the last section only apply on the top level. When we are matching *patterns within patterns*, these equivalence rules do not apply and we check for an exact text match, **excluding** alpha equivalence. This means that the following two channels are *not* equivalent:
+
+
+* `@for(  @for( @{ Nil | x } <- @Nil ){ Nil } <- @Nil ){ Nil }}`
+* `@for(  @for( @{ x | Nil } <- @Nil ){ Nil } <- @Nil ){ Nil }}`
+
+Note that while the patterns `@{ x | Nil }}` and `@{ Nil | x }` are equivalent at the top level, the patterns
+
+* `@for( @{ Nil | x } <- @Nil ){ z }`
+* `@for( @{ x | Nil } <- @Nil ){ z }`
+
+are *not*. When we are matching *patterns within patterns*, these equivalence rules do not apply and we check for an exact text match, up to alpha equivalence. The principle is that you can only look through the looking glass once: when checking name equivalence or matching a pattern to a name or a process, if a given pattern is part of a larger pattern, there has to be an exact match.
+
+Furthermore, we can't bind variables to parts of patterns. For example, the following send/receive will not match:
+
+    1 for( @ for( @{x!(y)} <- @Nil ){ Nil } <- @Nil ){ Nil }  |
+    2 @Nil!( for( @{x!(10)} <- @Nil ){ Nil } )
+
+Na&iuml;vely, one might expect these to match, binding `y` to `10`, but to match with the above receive, one must send something alpha equivalent to:
+
+    @Nil!( for( @x!(y) <- @Nil ){ Nil } )
+
+## Precedence Rules
+When writing more complicated patterns, one should be aware of precedence rules of the operators in Rholang. Take for example the following pattern:
+
+    @{@x!(Nil) | y!(Nil)}
+
+A priori, it is unclear whether this should be interpreted as `@{{@x!(Nil)} | {y!(Nil)}}`, where `x` is a process variable and `y` is a name variable, or as `@{@{x!(Nil) | y}!(Nil)}}`, where `x` is a name variable and `y` is a process variable. The chosen interpretation is vital, since the way we use `x` and `y` in the body depend on the type (process or name) of each of these terms. If the first interpretation is correct,
+
+    for( @{@x!(Nil) | y!(Nil)} <- @10 ){ @x!("success") | y!("success") }
+
+will compile, while
+
+    for( @{@x!(Nil) | y!(Nil)} <- @10 ){ x!("success") | @y!("success") }
+
+will not. If the second interpretation is correct, the opposite is true.
```

Comment:
> the opposite is true

From my read, if the second interpretation is correct, both will not compile.

Reply:
In practice parens are needed for this to compile. I just made some changes so it actually will compile.

Reply:
No I'm saying if the second interpretation is correct.

Reply:
Got it and fix looks good.